src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1355 @@
+            userCfg_.ap_.ssid_, userCfg_.ap_.pass_);
+    }
+
+    // Initialize mDNS.
+    ESP_ERROR_CHECK(mdns_init());
+    mdns_hostname_set(hostname_);
+    mdns_instance_name_set(hostname_);
+}
+
+//
+// EspIdfWiFi::init_softap()
+//
+void EspIdfWiFi::init_softap(std::string ssid, std::string pass)
+{
+    apIface_ = esp_netif_create_default_wifi_ap();
+    esp_netif_set_hostname(apIface_, hostname_);
+
+    wifi_config_t conf;
+    memset(&conf, 0, sizeof(wifi_config_t));
+
+    conf.ap.max_connection = 4;
+    conf.ap.beacon_interval = 100;
+    conf.ap.ssid_len = std::min(ssid.size(), (size_t)MAX_SSID_LEN);
+    memcpy(conf.ap.ssid, ssid.c_str(), conf.ap.ssid_len);
+
+    size_t pass_len = std::min(pass.size(), (size_t)MAX_PASSPHRASE_LEN);
+    if (pass_len == 0)
+    {
+        conf.ap.authmode = WIFI_AUTH_OPEN;
+    }
+    else
+    {
+        memcpy(conf.ap.password, pass.c_str(), pass_len);
+        conf.ap.authmode = WIFI_AUTH_WPA_WPA2_PSK;
+    }
+
+    ESP_ERROR_CHECK(esp_wifi_set_config(WIFI_IF_AP, &conf));
+}
+
+//
+// EspIdfWiFi::init_sta()
+//
+void EspIdfWiFi::init_sta()
+{
+    staIface_ = esp_netif_create_default_wifi_sta();
+    esp_netif_set_hostname(staIface_, hostname_);
+    ESP_ERROR_CHECK(esp_netif_dhcpc_start(staIface_));
+}
+
+//
+// EspIdfWiFi::sta_connect()
+//
+void EspIdfWiFi::sta_connect(
+    std::string ssid, std::string pass, uint8_t authmode, uint8_t channel)
+{
+    wifi_config_t conf;
+    memset(&conf, 0, sizeof(wifi_config_t));
+
+    size_t ssid_len = std::min(ssid.size(), (size_t)MAX_SSID_LEN);
+    memcpy(conf.sta.ssid, ssid.c_str(), ssid_len);
+
+    size_t pass_len = std::min(pass.size(), (size_t)MAX_PASSPHRASE_LEN);
+    memcpy(conf.sta.password, pass.c_str(), pass_len);
+    staConnectPass_ = std::move(pass);
+
+    LOG(INFO, "wifi: STA SSID: %s, PASS: %s", conf.sta.ssid, conf.sta.password);
+    if (pass_len)
+    {
+        conf.sta.threshold.authmode = (wifi_auth_mode_t)authmode;
+    }
+    else
+    {
+        conf.sta.threshold.authmode = WIFI_AUTH_OPEN;
+    }
+    conf.sta.pmf_cfg.required = false;
+    conf.sta.scan_method = WIFI_FAST_SCAN;
+    conf.sta.sort_method = WIFI_CONNECT_AP_BY_SIGNAL;
+    conf.sta.threshold.rssi = -100;
+    conf.sta.channel = channel;
+
+    ESP_ERROR_CHECK(esp_wifi_set_config(WIFI_IF_STA, &conf));
+    esp_wifi_connect();
+}
+
+
+//
+// EspIdfWiFi::last_sta_update()
+//
+void EspIdfWiFi::last_sta_update(
+    std::string ssid, std::string pass, uint8_t authmode, uint8_t channel)
+{
+    if (ssid != privCfg_.last_.ssid_ || pass != privCfg_.last_.pass_ ||
+        authmode != privCfg_.last_.sec_|| channel != privCfg_.channelLast_)
+    {
+        LOG(INFO, "wifi: Update last STA.");
+        strncpy(privCfg_.last_.ssid_, ssid.c_str(), MAX_SSID_SIZE);
+        privCfg_.last_.ssid_[MAX_SSID_SIZE] = '\0';
+        strncpy(privCfg_.last_.pass_, pass.c_str(), MAX_PASS_SIZE);
+        privCfg_.last_.pass_[MAX_PASS_SIZE] = '\0';
```

Comment:
reading this further there may be an issue here, if `MAX_PASS_SIZE` is 64 the actual field size is similar when passing down to underlying APIs. There is no null character present and instead we need to pass down password length in the structure IIRC

Reply:
Yes, I think it's covered so far.